### PR TITLE
OEP: distinguishing the central agent

### PR DESCRIPTION
**Pull request by kralicky**

### CODE REVIEW

File: enhancements/gateway/20220130-distinguishing-central-agent.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+1. An octet key pair is generated by cert-manager and mounted into both the agent and gateway pods
+2. The location of this keypair on disk is configured in both the agent and gateway config file (or checked for in a well-known location)
+3. The presence of the keypair triggers the agent to assert the relevant attribute.
+
+Configuration could be either done via config files or environment variables, or both.
+
+## Acceptance criteria:
+- [ ] The central agent can be distinguished from other agents using a well-known label in its session attributes list, obtained from the management API.
+- [ ] An agent with an ID that is not equal to the gateway's ID can successfully be set as the central agent.
+- [ ] Two agents can simultaneously be labeled as central agents.
```
Comment (dbason) [inline enhancements/gateway/20220130-distinguishing-central-agent.md:71]: What is the use case for having two agents as central agents?  Wouldn't we expect there to be one and only one?

Comment (kralicky) [reply enhancements/gateway/20220130-distinguishing-central-agent.md:71]: In this context "two agents" just means two agents with the same connection attribute. Could happen if we need to scale the agent or something similar.

**Review by alexandreLamarre — APPROVED** [review]
LGTM

**Review by dbason — APPROVED** [review]

File: enhancements/gateway/20220130-distinguishing-central-agent.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+## Summary:
+This proposal describes a mechanism by which the agent running in the central opni cluster can be distinguished from other clusters.
+
+## Use case:
+This is a prerequisite for several other features. For example, identifying the central opni cluster would be required for implementing a global metrics tenant.
+
+## Benefits:
+The ability to distinguish the central cluster from other clusters provides benefits such as:
+- It allows establishing an increased level of trust with a specific agent, allowing it to be authorized for tasks that cannot be entrusted to other agents.
+- It allows the UI, CLI, or other API consumers to distinguish the central cluster from other clusters, which is helpful for users.
```
Comment (sanjay920) [inline enhancements/gateway/20220130-distinguishing-central-agent.md:12]: This proposal should include all UI enhancements that can come out of this feature. The one that is most obvious is that the UI distinguishes the Opni cluster's agent in the UI. Anything else?
@codyrancher Can we add an estimate for this to the level of effort section ?

File: enhancements/gateway/20220130-distinguishing-central-agent.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+Any dependencies that must be met before the proposal can be implemented. This could include other projects, external systems, or third-party services that must be in place before the proposal can be completed.
+
+## Risks and contingencies:
+- If we do end up needing to know if a disconnected agent was labeled as the central agent, we can add session attributes to the "last known session details" metadata that is already tracked by the gateway for each agent connection. Will leave this feature out for now, since it could be confusing/redundant to have the attributes be available in two separate places.
+
+## Level of Effort:
+1 week total:
+Development: 2 days
+Testing: 2 days
+Docs: 1 day
```
Comment (sanjay920) [inline enhancements/gateway/20220130-distinguishing-central-agent.md:86]: see above

**Review by sanjay920 — CHANGES_REQUESTED** [review]

File: enhancements/gateway/20220130-distinguishing-central-agent.md (hunk 2 above)
Comment (kralicky) [reply enhancements/gateway/20220130-distinguishing-central-agent.md:12]: The only planned UI change will be adding some sort of distinguishing mark or label to the central cluster.

Comment (sanjay920) [reply enhancements/gateway/20220130-distinguishing-central-agent.md:12]: We discussed with Cody. UI will include a bubble to distinguish local agent

**Review by sanjay920 — APPROVED** [review]

**Pull request merged**